aiida_mlip/data/model.py

```diff
@@ -0,0 +1,224 @@
+"""Define Model Data type in AiiDA."""
+
+import hashlib
+from pathlib import Path
+from typing import Any, Optional, Union
+from urllib import request
+from urllib.parse import urlparse
+
+from aiida.orm import SinglefileData
+
+
+class ModelData(SinglefileData):
+    """Class to save a model file as an AiiDA data type.
+
+    The file can be a file that is stored locally or a new file to download.
+    """
+
+    @staticmethod
+    def _calculate_hash(file: Union[str, Path]) -> str:
+        """Calculate the hash of a file.
+
+        Parameters
+        ----------
+        file : Union[str, Path]
+            Path to the file for which hash needs to be calculated.
+
+        Returns
+        -------
+        str
+            The SHA-256 hash of the file.
+        """
+        # Calculate hash
+        buf_size = 65536  # reading 64kB (arbitrary) at a time
+        sha256 = hashlib.sha256()
+        with open(file, "rb") as f:
+            # calculating sha in chunks rather than 1 large pass
+            while data := f.read(buf_size):
+                sha256.update(data)
+        file_hash = sha256.hexdigest()
+        return file_hash
+
+    @staticmethod
+    def _check_existing_file(
+        file: Union[str, Path],
+    ) -> Path:
+        """Check if a file already exists and return the path of the existing file if it does.
+
+        Parameters
+        ----------
+        file : Union[str, Path]
+            Path to the downloaded model file.
+
+        Returns
+        -------
+        Path
+            The path of the model file of interest (same as input path if no duplicates were found).
+        """
+        file_hash = ModelData._calculate_hash(file)
+
+        def is_diff_file(curr_path: Path):
+            return curr_path.is_file() and not curr_path.samefile(file)
+
+        file_folder = Path(file).parent
+        for existing_file in filter(is_diff_file, file_folder.rglob("*")):
+            if ModelData._calculate_hash(existing_file) == file_hash:
```

Comment:
Would it make sense to make this method into a class method and call this via `cls._calculate_hash`?

My feeling is converting it would be slightly preferable, but I'm not sure if it makes a difference.